tests/ui/impl-trait/unsized_coercion.rs

```diff
@@ -0,0 +1,21 @@
+//! This test checks that opaque types get unsized instead of
+//! constraining their hidden type to a trait object.
+
+//@ revisions: next old
+//@[next] compile-flags: -Znext-solver
+//@[old] check-pass
```

Comment:
I explained why this fails with `-Znext-solver` in https://github.com/rust-lang/rust/pull/125610#issuecomment-2133998304

I feel like this should compile. We could fix this by partially reverting https://github.com/rust-lang/rust/pull/119989:
- store `sub_relations` in the `FnCtxt` instead of the `InferCtxt` (so they aren't used by anything inside of the trait solver
- return them as external constraints inside of the solver

This allows us to consider subtype relations in `obligations_for_self_ty`, e.g. when coercing types, which seems a lot more clearly desirable to me, without having to pass them into canonical queries. The reason I removed them is

> we don't track them when canonicalizing or when freshening,
resulting in instable caching in the old solver, and issues when
instantiating query responses in the new one.

Tracking them in canonical inputs has a significant performance and complexity cost and is unnecessary for this case.

This is very similar to the following which has previously never worked:
```rust
fn hello() {
    let y = Default::default();
    let x = Box::new(y);
    let z: Box<dyn Send> = x;
    //~^ the size for values of type `dyn Send` cannot be known at compilation time
    let _: i32 = y;
}
```